Mathlib/NumberTheory/LSeries/Convolution.lean

```diff
@@ -0,0 +1,203 @@
+Technically, this is done by transporting the existing definition for `ArithmeticFunction R`;
+see `LSeries.convolution`. We show that these definitions agree (`LSeries.convolution_def`).
+
+We then consider the case `R = ℂ` and show that `L (f ⍟ g) = L f * L g` on the common domain
+of convergence of the L-series `L f`  and `L g` of `f` and `g`; see `LSeries_convolution`
+and `LSeries_convolution'`.
+-/
+
+open scoped LSeries.notation
+
+open Complex LSeries
+
+/-!
+### Dirichlet convolution of two functions
+-/
+
+open BigOperators
+
+/-- We turn any function `ℕ → R` into an `ArithmeticFunction R` by setting its value at `0`
+to be zero. -/
+def toArithmeticFunction {R : Type*} [Zero R] (f : ℕ → R) : ArithmeticFunction R where
+  toFun n := if n = 0 then 0 else f n
+  map_zero' := rfl
+
+/-- If we consider an arithmetic function just as a function and turn it back into an
+arithmetic function, it is the same as before. -/
+@[simp]
+lemma ArithmeticFunction.toArithmeticFunction_eq_self {R : Type*} [Zero R]
+    (f : ArithmeticFunction R) :
+    toArithmeticFunction f = f := by
+  ext n
+  simp (config := {contextual := true}) [toArithmeticFunction, ArithmeticFunction.map_zero]
+
+/-- Dirichlet convolution of two sequences.
+
+We define this in terms of the already existing definition for arithmetic functions. -/
+noncomputable def LSeries.convolution {R : Type*} [Semiring R] (f g : ℕ → R) : ℕ → R :=
+  (toArithmeticFunction f * toArithmeticFunction g :)
+
+@[inherit_doc]
+scoped[LSeries.notation] infixl:70 " ⍟ " => LSeries.convolution
+
+open scoped LSeries.notation
+
+/-- The product of two arithmetic functions defines the same function as the Dirichlet convolution
+of the functions defined by them. -/
+lemma ArithmeticFunction.mul_to_convolution {R : Type*} [Semiring R] (f g : ArithmeticFunction R) :
+    f ⍟ g = (f * g :) := by
+  simp only [convolution, ArithmeticFunction.toArithmeticFunction_eq_self]
+
+namespace LSeries
+
+lemma convolution_def {R : Type*} [Semiring R] (f g : ℕ → R) :
+    f ⍟ g = fun n ↦ ∑ p in n.divisorsAntidiagonal, f p.1 * g p.2 := by
+  ext n
+  simp only [convolution, toArithmeticFunction, ArithmeticFunction.mul_apply,
+    ArithmeticFunction.coe_mk, mul_ite, mul_zero, ite_mul, zero_mul]
+  refine Finset.sum_congr rfl fun p hp ↦ ?_
+  obtain ⟨hp₁, hp₂⟩ := Nat.mem_divisorsAntidiagonal.mp hp
+  obtain ⟨h₁, h₂⟩ := mul_ne_zero_iff.mp (hp₁.symm ▸ hp₂)
+  simp only [h₂, ↓reduceIte, h₁]
+
+@[simp]
+lemma convolution_map_zero {R : Type*} [Semiring R] (f g : ℕ → R) : (f ⍟ g) 0 = 0 := by
+  simp only [convolution_def, Nat.divisorsAntidiagonal_zero, Finset.sum_empty]
+
+
+/-!
+### Multiplication of L-series
+-/
+
+open Set Nat in
+/-- We give an expression of the `LSeries.term` of the convolution of two functions. -/
+lemma term_convolution (f g : ℕ → ℂ) (s : ℂ) (n : ℕ) :
+    term (f ⍟ g) s n =
+      ∑' (b : (fun p : ℕ × ℕ ↦ p.1 * p.2) ⁻¹' {n}),
+        term f s b.val.1 * term g s b.val.2 := by
```

Comment:
I was confused by this – the formulation seemed bizarre ("why is he taking a `tsum` over an obviously finite set?") until I realised that the index set on the RHS is infinite if `n = 0` (in particular, it doesn't match `Nat.divisorsAntidiagonal n`). Maybe a clarifying comment might be a good idea? I also think it might be a good idea to also provide the formulation in terms of `Nat.divisorsAntidiagonal` first, as a separate lemma, and then deduce this formulation from that (the total length of those two proofs would be comparable to this single big one).

Reply:
OK. I'm splitting off the `Nat.divisorsAntidiagonal` version. I have also added some text to the docstrings and golfed the proof a little bit.